enhancements/cloud-integration/azure/azure-workload-identity.md

```diff
@@ -0,0 +1,358 @@
+`ServiceAccount` referenced by a pod being created, environment variables are set for the pod for the `AZURE_CLIENT_ID`, `AZURE_TENANT_ID` and `AZURE_FEDERATED_TOKEN_FILE`.
+
+This will be similar to how CCO deploys the [AWS Pod Identity webhook](https://github.com/openshift/aws-pod-identity-webhook) which we have forked for use by user workloads.
+
+#### Variation [optional]
+
+TBD
+
+### API Extensions
+
+None as of now.
+
+### Implementation Details/Notes/Constraints [optional]
+
+TBD
+
+### Risks and Mitigations
+
+- The feature this work relies on was recently made public preview. What is the timeline for GA for Workload identity federation support for Managed Identities?
+- How will security be reviewed and by whom?
+- How will UX be reviewed and by whom?
+
+### Drawbacks
+
+The pod identity webhook deployed for AWS has received little ongoing maintenance since its initial deployment by CCO and this proposal adds yet another webhook to be lifecycled by CCO, however upstream seems to be moving in this direction for providing client details as opposed to config secrets. It is likely best for compatibility with how operators currently obtain client information from a
+config secret while also respecting the environment variables that would be set by the webhook. Additionally, upstream projects may reject the notion of reading these details from a config secret but that has yet to be seen.
+
+## Design Details
+
+### Open Questions [optional]
+
+- From where should CCO source the mutating admission webhook for deployment? In order to generate our own build of the image backing the webhook we would have to fork [Azure/azure-workload-identity](https://github.com/Azure/azure-workload-identity)([dockerfile](https://github.com/Azure/azure-workload-identity/blob/main/docker/webhook.Dockerfile)).
+
+### Test Plan
+
+An e2e test job will be created similar to the [e2e-gcp-manual-oidc](https://github.com/openshift/release/pull/22552) that,
+- Extracts `CredentialsRequests` from the release image.
+- Processes `CredentialsRequests` with `ccoctl` to generate secret and `Authentication` configuration manifests.
+- Moves the generated manifests into the manifests directory used for install.
+- Runs the normal e2e suite against the resultant cluster.
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+- Ability to utilize the enhancement end to end
+- End user documentation, relative API stability
+- Sufficient test coverage
+- Gather feedback from users rather than just developers
+- Enumerate service level indicators (SLIs), expose SLIs as metrics
+- Write symptoms-based alerts for the component(s)
+
+#### Tech Preview -> GA
```

Comment:
I would like to see the feature introduced as gated under TechPreviewNoUpgrade https://github.com/openshift/api/blob/master/config/v1/types_feature.go#L111 and then promoted once it is demonstrated to be working reliably.  There are existing CI jobs running installation on Azure

Reply:
There is no central CCO function being performed where we could check a feature gate so the feature gate will need to be checked in every operator in context of what the operator is doing to create an Azure client.

Reply:
I'm not yet sure if every operator is aware of feature gates but I expect this will be a lot of work to implement and plumb through each operator.

Reply:
Here's my proposal which I believe achieves David's intent while balancing the effort to plumb this through to a dozen operators.

- Operators add support for the new config secret, unencumbered by the feature gate.
- prior to promotion to GA ccoctl adds a manifest setting TechPreviewNoUpgrade feature-set in the bundle of manifests it creates
- CCO requires TechPreviewNoUpgrade feature gate to deploy the webhook.

Once deemed GA, ccoctl stops annotating config secrets. CCO deploys the webhook w/o the featuregate. Installer retains the manifest validation for potential future use.

If that is not robust enough we could pivot to `ccoctl` adding a `installer.openshift.io/feature-set=TechPreviewNoUpgrade` annotation to each config secret and the installer then refuses to render Ignition assets whenever any manifest has that annotation but the Installer's existing TechPreviewNoUpgrade [feature-set](https://github.com/openshift/enhancements/blob/master/enhancements/installer/feature-sets.md) is unset.

Reply:
> There is no central CCO function being performed where we could check a feature gate so the feature gate will need to be checked in every operator in context of what the operator is doing to create an Azure client.

This change is already going to require changes to such operators.  Gating something delivered across multiple operators is exactly what such gating is for.  Getting the gate added and teaching those operators how to read a gate while consuming it is very reasonable. 

> I'm not yet sure if every operator is aware of feature gates but I expect this will be a lot of work to implement and plumb through each operator.

Compared to the rest of the work, requiring a gate is minor.

Reply:
Just so we're sure we're all talking about the same thing, this is vendoring openshift/api and writing

```go
	if sets.NewString(configv1.FeatureSets[featureSet].Enabled...).Has("CoolNewFeature") {
		// do new thing
	} else {
		// do old thing
	}
```

working example in another operator here: https://github.com/openshift/cluster-kube-apiserver-operator/blob/0eb5517f7faec395d53258c590ba0936d8536d3d/pkg/cmd/render/render.go#L307-L315

Reply:
@deads2k That makes sense. Would the first step be to add a feature to the `TechPreviewNoUpgrade` list like this https://github.com/openshift/api/pull/1436?

Reply:
Marking this as resolved since it's been sorted out offline. PRs to soon be reviewed and merged.